test: verify process instance modification in multi instance subprocesses using ancestor selection

mustafadagher opened the pull request:
## Description

This PR refactors an existing test and adds three new test cases for process instance modification within multi-instance ad-hoc subprocesses. The changes improve test clarity by explicitly retrieving multi-instance subprocess keys rather than deriving them from task flow scopes.

- Refactored `shouldUseAncestorSelectionInsideMultiInstances` to explicitly retrieve multi-instance subprocess instances
- Added comprehensive test coverage for ancestor selection in ad-hoc subprocess scenarios

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [ ] Enable backports when necessary (fex. [for bug fixes](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#backporting-changes) or [for CI changes](https://github.com/camunda/camunda/wiki/CI-&-Automation#when-to-backport-ci-changes)).

## Related issues

closes #40226

mustafadagher committed:
test: verify modification possible for element inside MI ad-hoc subprocess

if a single instance of the MI body is selected as the `ancestorElementInstanceKey`

mustafadagher committed:
test: refactor the multi-instance subprocess modification test to clarify usage of inner instance as ancestor scope

mustafadagher committed:
test: should modify elements in a nested multi-instance subprocesses using ancestor selection

korthout commented on zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java:
```diff
@@ -1420,8 +1437,294 @@ public void shouldUseAncestorSelectionInsideMultiInstances() {
+            .create();
+
+    // Get all multi-instance ad-hoc subprocess instances
+    final var miInnerInstances =
```
🔧This name implies it's the inner instance of the multi-instance, but it's not. It is the inner instance of the ad hoc sub process instance.

korthout commented on zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java:
```diff
@@ -1420,8 +1437,294 @@ public void shouldUseAncestorSelectionInsideMultiInstances() {
+        .withInstanceKey(processInstanceKey)
+        .modification()
+        .terminateElement(aTasks.getFirst().getKey())
+        .activateElement("B", miInnerInstances.getFirst().getKey())
```
🔧Here you are choosing the first ad hoc sub process' inner instance, assuming that it's the same as the flow scope of the task instance that you're terminating. Luckily that's the case, because the ordering is like that on the log. However, it would be better to be more explicit about the activation. What we want to do is move A to B within it's flowscope. Like you say in the comment "terminate task A in the first ad-hoc instance and activate task B there", so let's choose to activate B inside the same flowscope explicitly:
```suggestion
        .activateElement("B", aTasks.getFirst().getValue().getFlowScopeKey())
```

korthout commented on zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java:
```diff
@@ -1420,8 +1437,294 @@ public void shouldUseAncestorSelectionInsideMultiInstances() {
+        .extracting(Record::getValue)
+        .extracting(ProcessInstanceRecordValue::getFlowScopeKey)
+        .describedAs("Expect B Task to be activated in the specific ad-hoc subprocess instance")
+        .contains(miInnerInstances.getFirst().getKey());
```
🔧Here we can argue the same as before. We want to assert that Task B was activated in the same ad-hoc subprocess inner instance as Task A. So let's just check that the flowscope is the same as the flow scope of Task A. 

Taking this step in addition to my previous suggestion removes the need for the miInnerInstances variable, which means one fewer RecordingExporter usage. So the test would be slightly faster as well.

mustafadagher replied:
>Taking this step in addition to my previous suggestion removes the need for the miInnerInstances variable, which means one fewer RecordingExporter usage. So the test would be slightly faster as well.

I actually refactored the above test that only relied on querying the flowscope key of the task to instead use the mi-instance key, to make the use case more clear.
I'm not concerned personally about the few milliseconds the extra RecordingExporter query adds, trading off clarity for it.

I'll add an extra assert that verifies that task-a flowscope is the same as the first adhoc-inner-instance key.

korthout reviewed (changes requested):
Nice work @mustafadagher 👏 

🔧 I have a few suggestions, and ❌ two change requests.

Please have a look.

korthout commented on zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java:
```diff
@@ -1420,8 +1437,294 @@ public void shouldUseAncestorSelectionInsideMultiInstances() {
+  }
+
+  @Test
+  public void shouldModifyElementsInMultiInstanceAdHocSubprocessWithAiAgentTaskPattern() {
```
>Perhaps if it would target different ancestor scopes (one targets an ad hoc subprocess inner instance, and the other target the ad hoc subprocess itself).

That might be a different edge case

mustafadagher committed:
test: verify new ad-hoc inner instance creation with MI subprocess ancestor

Add test case to verify that when modifying a process instance with a
multi-instance ad-hoc subprocess, activating an element using the ad-hoc
subprocess instance (not the inner instance) as the ancestor correctly
creates a new ad-hoc subprocess inner instance within that MI instance.

The test ensures that the modification command properly handles the
ancestry hierarchy and creates the appropriate nesting structure when
the specified ancestor is at the ad-hoc subprocess level rather than
at the inner instance level.

korthout reviewed (approved):
Thanks for the changes @mustafadagher 🚀 

LGTM 👍
